www/content/workstation/upgrade_labs.md

```diff
@@ -0,0 +1,330 @@
+cannot be saved upstream, though they can still be uploaded
+to a Chef Server:
+
+  - logrotate (v1.9.2)
+  - windows (v1.44.1)
+  - chef_handler (v1.4.0)
+
+You're ready to begin!
+
+Start with 'kitchen converge'.  As you identify issues, you
+can modify cookbooks in their original checkout locations or
+in the repository's cookbooks directory and they will be picked
+up on subsequent runs of 'kitchen converge'.
+```
+
+## Converge Locally
+
+{{< note >}}
+Be cautious when running your cookbooks locally. It is possible to accidentally impact production infrastructure based on settings in your cookbooks and attributes. Read over your cookbooks and attributes, especially attributes being set in roles and environments.  If needed, override attributes to be appropriate for local testing in your `kitchen.yml`.
+{{< /note >}}
+
+### Increment the Chef Infra Client Version
+
+In the `kitchen.yml` file, change the `product_version` line to `16`:
+```yaml
+provisioner:
+  name: chef_zero
+  product_name: chef
+  product_version: 16  # Change this line
+```
+
+If needed, you can "step forward" by first going from 12 to 13, correcting issues, then 13 to 14, etc.
+
+### Attempt a Converge and Check for Errors
+
+Run, in the `node-MY_NODE-repo` directory:
+```
+$ chef exec kitchen converge
+```
+Watch for Chef Infra errors. If any occur, fix them.  Also consider running cookstyle (see below). Repeat as needed.
+
+### Run CookStyle
+To check for version upgrade issues, run:
+```
+$ chef exec cookstyle cookbooks/some-cookbook
+```
+Repeat this process for each cookbook that the node consumes.
+
+## Correct any Cookbook Style Issues
+### Using auto-correct
+To auto-correct cookbook issues, run:
+```
+$ chef exec cookstyle -a cookbooks/some-cookbook
+```
+Other issues may require manual intervention and editing.
+Repeat this process for each cookbook that the node consumes.
+
+### Using the VSCode Plugin
+While any editor can be used, the Chef Infra extension for Visual Studio Code features several code generators and helpful features, such as running cookstyle when recipes are saved.
+
+### Check for Data Bags
+If data bags are used, you will need a `data_bags` directory in your repo.
+Pull down the data_bags by running:
+```
+$ chef exec knife download data_bags
+```
+
+### Check for Server Searches
+Check your cookbook code for Chef Infra Server searches, which will not be possible in an Effortless context. Identify locations making search calls and replace with other mechanisms of service discovery.
+```
+$ grep 'search(' -rn cookbooks
+```
+
+### Commit Your Changes to the Cookbooks
+
+As you make changes to the cookbooks, follow normal SDLC practices by committing your changes to your cookbooks and submitting your changes to your cookbook pipeline to be tested by your automated testing system. Once the changes have passed testing, they cookbooks should receive new version numbers and be published to the new chef server by the continuous deployment system.
```

Comment:
When we went to upload cookbooks, we hit this error

```
 jobrien@joshs-mbp  ~/.chef/node-node-01-repo   master  knife upload cookbooks
ERROR: Attempt to use relative path 'cookbooks' when current directory is outside the repository path.
ERROR: Current working directory is '/Users/jobrien/.chef/node-node-01-repo'.
 ✘ jobrien@joshs-mbp  ~/.chef/node-node-01-repo   master  knife upload cookbooks --profile new-server
ERROR: Attempt to use relative path 'cookbooks' when current directory is outside the repository path.
ERROR: Current working directory is '/Users/jobrien/.chef/node-node-01-repo'.
 ✘ jobrien@joshs-mbp  ~/.chef/node-node-01-repo   master  knife upload ~/.chef/node-node-01-repo/cookbooks --profile new-server
/opt/chef-workstation/embedded/lib/ruby/gems/2.7.0/gems/chef-15.10.12/lib/chef/resource.rb:1384: warning: Using the last argument as keyword parameters is deprecated; maybe ** should be added to the call
/opt/chef-workstation/embedded/lib/ruby/gems/2.7.0/gems/chef-15.10.12/lib/chef/node_map.rb:58: warning: The called method `set' is defined here
ERROR: Chef::ChefFS::FileSystem::OperationNotAllowedError: cannot have a child created under it
```

Reply:
Adding `--chef-repo-path .` fixed the path issue for me. I did not encounter the child creation issue, with invocation
```
chef exec knife upload cookbooks --profile new-server --chef-repo-path . --key ../keys/new-janedoe.pem
```